ddl/ddl_api.go

```diff
@@ -636,6 +637,117 @@ func checkConstraintNames(constraints []*ast.Constraint) error {
+	if meta.Partition.Type == model.PartitionTypeRange {
+		newPars := part.Definitions
+		set := make(map[string]bool)
+		for _, newPar := range newPars {
+			if _, ok := set[strings.ToLower(newPar.Name)]; ok {
+				return ErrSameNamePartition.GenByArgs(newPar.Name)
+			}
+			set[strings.ToLower(newPar.Name)] = true
+		}
+	}
+	return nil
+}
+
+// checkCreatePartitionValue values less than value must be strictly increasing for each partition.
+func checkCreatePartitionValue(meta *model.TableInfo, part *model.PartitionInfo) error {
+	if meta.Partition.Type == model.PartitionTypeRange {
+		newDefs := part.Definitions
+		rangeValue := newDefs[0].LessThan[0]
+		ifMaxvalue := strings.EqualFold(rangeValue, "MAXVALUE")
+		if ifMaxvalue && len(newDefs) > 1 {
+			return errors.Trace(ErrPartitionMaxvalue)
+		} else if ifMaxvalue && len(newDefs) == 1 {
+			return nil
+		}
+		currentRangeValue, err := strconv.Atoi(rangeValue)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		for i := 1; i < len(newDefs); i++ {
+			ifMaxvalue := strings.EqualFold(newDefs[i].LessThan[0], "MAXVALUE")
+			if ifMaxvalue && i == len(newDefs)-1 {
+				return nil
+			} else if ifMaxvalue && i != len(newDefs)-1 {
+				return errors.Trace(ErrPartitionMaxvalue)
+			}
+
+			nextRangeValue, err := strconv.Atoi(newDefs[i].LessThan[0])
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if nextRangeValue <= currentRangeValue {
+				return errors.Trace(ErrRangeNotIncreasing)
+			}
+			currentRangeValue = nextRangeValue
+		}
+	}
+	return nil
+}
+
+func buildCreateTablePartitionInfo(ctx sessionctx.Context, d *ddl, tbInfo *model.TableInfo, s *ast.CreateTableStmt, cols []*table.Column) error {
+	if s.Partition == nil {
+		return nil
+	}
+	pi := &model.PartitionInfo{
+		Type:   s.Partition.Tp,
+		Enable: ctx.GetSessionVars().EnableTablePartition,
+	}
+	if s.Partition.Expr != nil {
+		buf := new(bytes.Buffer)
+		s.Partition.Expr.Format(buf)
+		pi.Expr = buf.String()
+		if s.Partition.Tp == model.PartitionTypeRange {
+			if _, ok := s.Partition.Expr.(*ast.ColumnNameExpr); ok {
+				for _, col := range cols {
+					name := strings.Replace(col.Name.String(), ".", "`.`", -1)
+					if !(col.Tp == mysql.TypeLong || col.Tp == mysql.TypeLonglong) && fmt.Sprintf("`%s`", name) == pi.Expr {
+						return errors.Trace(ErrNotAllowedTypeInPartition.GenByArgs(pi.Expr))
+					}
+				}
+			}
+		}
+	} else if s.Partition.ColumnNames != nil {
+		pi.Columns = make([]model.CIStr, 0, len(s.Partition.ColumnNames))
+		for _, cn := range s.Partition.ColumnNames {
+			pi.Columns = append(pi.Columns, cn.Name)
+		}
+	}
+	for _, def := range s.Partition.Definitions {
+		// TODO: generate multiple global ID for paritions.
+		pid, err1 := d.genGlobalID()
+		if err1 != nil {
+			return errors.Trace(err1)
+		}
+		piDef := model.PartitionDefinition{
+			Name:    def.Name,
+			ID:      pid,
+			Comment: def.Comment,
+		}
+		for _, expr := range def.LessThan {
+			buf := new(bytes.Buffer)
```

Comment:
move it out the for-loop and reuse.
Note that `buf.String()` will make a new copy, so you don't need to worry about the memory reuse.